src/atomic.rs

```diff
@@ -1,160 +1,380 @@
-use cid::multihash::Code::Blake2b256;
-use cid::multihash::MultihashDigest;
+#![allow(dead_code)] // TODO: Remove
+
+use cid::multihash::{Blake2b256, MultihashDigest};
+use cid::multihash::{Code, Hasher};
+use fvm_ipld_hamt::BytesKey;
+use ipc_gateway::IPCAddress;
+use std::cell::Cell;
+use std::collections::HashMap;
+use std::ops::Deref;
+
 use cid::Cid;
-use fil_actors_runtime::cbor;
+use fvm_ipld_blockstore::Blockstore;
 use fvm_ipld_encoding::tuple::Deserialize_tuple;
-use fvm_ipld_encoding::{serde_bytes, Cbor, RawBytes, DAG_CBOR};
-use fvm_shared::MethodNum;
+use fvm_ipld_encoding::{Cbor, CborStore, RawBytes, DAG_CBOR};
 use primitives::{TCid, THamt};
-use serde::{de::DeserializeOwned, Serialize};
-
-/// Trait that determines the functions that need to be implemented by
-/// a state object to be lockable and be used in an atomic execution.
-///
-/// Different strategies may be used to merge different locked state to
-/// prepare the actor state for the execution, and for the merging of the
-/// output of the execution to the original state of the actor.
-pub trait MergeableState<S: Serialize + DeserializeOwned> {
-    /// Merge a locked state (not necessarily the output) to the current state.
-    fn merge(&mut self, other: Self) -> anyhow::Result<()>;
-    /// Merge the output of an execution to the current state.
-    fn merge_output(&mut self, other: Self) -> anyhow::Result<()>;
+use serde::Deserialize;
+use serde::{self, de::DeserializeOwned, Serialize};
+use serde_tuple::Serialize_tuple;
+
+pub trait LockableState: Cbor {
+    fn lock(&mut self) -> anyhow::Result<()>;
+    fn unlock(&mut self) -> anyhow::Result<()>;
+    fn is_locked(&self) -> bool;
+    fn cid(&self) -> Cid {
+        cid_from_cbor(self)
+    }
 }
 
-/// Internal map kept by actor supporting atomic executions to track
-/// the states that have been locked and are used in an atomic exec.
-#[allow(dead_code)]
-pub type LockedMap<T> = TCid<THamt<Cid, LockableState<T>>>;
+fn cid_from_cbor(obj: &impl Cbor) -> Cid {
+    Cid::new_v1(
+        DAG_CBOR,
+        Code::Blake2b256.digest(&obj.marshal_cbor().unwrap()),
+    )
+}
 
-/// Trait that specifies the interface of an actor state able to support
-/// atomic executions.
-pub trait LockableActorState<T>
+/// State of an actor including a lock to support atomic executions.
+#[derive(Debug, Serialize_tuple, Deserialize_tuple)]
+pub struct InputState<T>
```

Comment:
`InputState` represents the whole state for the actor or a part of the whole state of the actor that can be locked individually for the atomic execution? (my question will probably be answered in your next PR :) but just in case)

Reply:
It is assumed to be a piece of the actor state.